src/NHibernate/Util/ExpressionsHelper.cs

```diff
@@ -526,7 +526,12 @@ private static IType TryGetComponentPropertyType(IAbstractComponentType componen
 				return persister != null;
 			}
 
-			return TryGetEntityPersister(convertedType, sessionFactory, out persister);
+			if (TryGetEntityPersister(convertedType, sessionFactory, out persister))
+				return true;
+
+			// Assume type conversion doesn't change entity type
+			persister = currentEntityPersister;
+			return true;
```

Comment:
HQL doesn't support entity casts. So to me this whole method looks like unnecessary complication.  At least we shouldn't loose already resolved persister.